#8534: Publish tt-metal docs to the central site

dimitri-tenstorrent opened the pull request:
### Ticket
[Link to Github Issue
](https://github.com/tenstorrent/tt-metal/issues/8534)

### Problem description
Create and push ttnn and tt-metalium documentation to docs-test repo for the centralized repository.
A particular challenge is to build the doxygen documentation for C++ as well as pick up the documentation from Python libraries of ttnn and tt-metalium that rely on the installation of the tt-metal wheel.

In order to capture Doxygen documentation, we run doxygen and add the resulting .xml files into the docs repo source so we can build documentation as part of the centralized docs repo build.

We also check-in all existing .rst files while not copying `conf.py` file so we can use an adapted conf.py in docs repo.

### What's changed
A new workflow is added that is triggered upon a release workflow.

### Checklist
- [x] Post commit CI passes
- [ ] Model regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

dimitri-tenstorrent committed:
#8534: Add the first iteration of the doc pipeline that brings the docs into the centralized repo
 #8534: Add packaging of TTNN Docs to the Deployment workflow

tt-rkim commented on .github/workflows/docs-deployment.yaml:
```diff
@@ -0,0 +1,98 @@
+        arch: [grayskull]
+    env:
+      TT_METAL_ENV: ${{ vars.TT_METAL_ENV }}
+      DOCS_VERSION: v0.50.0
```
May want to parameterize this here, may provide from an input?

tt-rkim commented on .github/workflows/docs-deployment.yaml:
```diff
@@ -0,0 +1,98 @@
+    strategy:
+      # Do not fail-fast because we need to ensure all tests go to completion
+      # so we try not to get hanging machines
+      fail-fast: false
```
Likely don't need `fail-fast` here for the reason listed, but I'm wondering if this has anything to do with `concurrency`. I'm also gonna guess so

tt-rkim commented on .github/workflows/docs-deployment.yaml:
```diff
@@ -0,0 +1,98 @@
+        run: tar -xvf ttm_${{ matrix.arch }}.tar
+      - uses: ./.github/actions/install-python-deps
+      - name: Build Doxygen Docs
+        timeout-minutes: 15
```
Did this hang before?

dimitri-tenstorrent replied:
Nope - removed.

tt-rkim commented on .github/workflows/docs-deployment.yaml:
```diff
@@ -0,0 +1,98 @@
+          git tag -a ttnn_$DOCS_VERSION -m "ttnn and tt-metalium documentation version $DOCS_VERSION"
+          git tag -a tt-metalium_$DOCS_VERSION -m "ttnn and tt-metalium documentation version $DOCS_VERSION"
+          echo "Pushing the new docs"
+          git push && git push --tags
```
POSIX new line

tt-rkim replied:
![Uploading Screenshot 2024-07-17 at 10.45.22 AM.png…]()

tt-rkim commented on .github/workflows/docs-deployment.yaml:
```diff
@@ -0,0 +1,98 @@
+          source ${{ github.workspace }}/python_env/bin/activate
+          cd $TT_METAL_HOME
+          export PYTHONPATH=$TT_METAL_HOME
+          doxygen
```
Perhaps create an issue to abstract away all these steps inside the `docs/Makefile`?

tt-rkim commented on .github/workflows/docs-deployment.yaml:
```diff
@@ -0,0 +1,98 @@
+          git config --global user.email "${GITHUB_USER}@tenstorrent.com"
+          git config --global user.name "${GITHUB_USER}"
+          echo "About to clone"
+          git clone https://${GITHUB_USER}:${{ secrets.TTBOT_DOCS_ACCESS }}@github.com/tenstorrent/docs-test.git
```
Hmm... I'm wondering if you could've used `actions/checkout@v4` with a specific token instead of this. Oh well, not blocking

dimitri-tenstorrent replied:
Can the action/checkout@v4 be also used to checkout another repo, not only the current one?

tt-rkim commented on .github/workflows/docs-deployment.yaml:
```diff
@@ -0,0 +1,98 @@
+          python update_tags.py tt-metalium $DOCS_VERSION
+          python update_tags.py ttnn $DOCS_VERSION
+
+          git add .
```
Hmm, so we did decide to build on the docs repo?

dimitri-tenstorrent replied:
Yeap. The issue is that if we build here and convert to .md some data would be missing. 
The alternative would be to build to .rst and then commit that into the repo. 
Let's explore if we could make it happen in the follow up issue (I'll create one ).

dimitri-tenstorrent replied:
https://github.com/tenstorrent/tt-metal/issues/10450

tt-rkim commented on .github/workflows/docs-deployment.yaml:
As a general comment, should we start tying the main docs site deployment with the current metal github pages workflow as well? Any reason to have a separate workflow like this?

dimitri-tenstorrent replied:
Good thought, I need to have a little more requirements here of how this would work. I ll create an issue.

dimitri-tenstorrent replied:
https://github.com/tenstorrent/tt-metal/issues/10449

dimitri-tenstorrent commented on .github/workflows/docs-deployment.yaml:
```diff
@@ -0,0 +1,98 @@
+      - uses: ./.github/actions/install-metal-dev-deps
+        with:
+          os: ubuntu-20.04
+      - uses: actions/download-artifact@v4
```
I am not sure if we are talking about the same things. Here is the code that ensures that there is an artifact.

```
  build-docs:
    needs: create-and-upload-draft-release
    uses: ./.github/workflows/docs-deployment.yaml
```

dimitri-tenstorrent commented:
@tt-rkim You have mentioned a few screenshots in your responses, but I can't see them.

dimitri-tenstorrent commented on .github/workflows/docs-deployment.yaml:
```diff
@@ -0,0 +1,94 @@
+name: "Release Docs: package the docs and publish to Docs the repo"
+
+on:
+  workflow_dispatch:
```
Do not allow dispatch. Create a wrapper flow.

dimitri-tenstorrent committed:
#8534: Enable the push to the tenstorrent.github.io repo
 Update TTNN Get started to edit the location of installation.html and the tutorials

dimitri-tenstorrent commented on .github/workflows/docs-release.yaml:
```diff
@@ -0,0 +1,115 @@
+          echo "Register new documentation versions"
+          python update_tags.py tt-metalium $DOCS_VERSION
+          python update_tags.py ttnn $DOCS_VERSION
+          git checkout -b test_v51_deploy
```
Do not forget to change.

dimitri-tenstorrent commented on .github/workflows/docs-release.yaml:
```diff
@@ -0,0 +1,115 @@
+          git tag -a ttnn_$DOCS_VERSION -m "ttnn and tt-metalium documentation version $DOCS_VERSION"
+          git tag -a tt-metalium_$DOCS_VERSION -m "ttnn and tt-metalium documentation version $DOCS_VERSION"
+          echo "Pushing the new docs"
+          git push -u origin test_v51_deploy && git push --tags
```
Do not forget to change.

dimitri-tenstorrent commented on .github/workflows/docs-latest-public-wrapper.yaml:
```diff
@@ -2,16 +2,21 @@ name: "[post-commit] Docs build and deploy to GitHub pages on main"
-  build-docker-artifact:
-    uses: ./.github/workflows/build-docker-artifact.yaml
-    secrets: inherit
+  # build-docker-artifact:
```
The changes to this file are purely for testing purposes. There should be no difference besides the release section at the end.

dimitri-tenstorrent committed:
#8534: Update the target to main now that we built the confidence with the system

dimitri-tenstorrent committed:
#8534: Revert the wrapper to previous value

tt-rkim commented on .github/workflows/docs-release.yaml:
```diff
@@ -0,0 +1,115 @@
+    env:
+      DOCS_VERSION: ${{ inputs.version }}
+      DOCS_REPO: tenstorrent.github.io
+      ARCH_NAME: ${{ matrix.arch }}
```
`ARCH_NAME` and `LD_LIBRARY_PATH` may actually be unnecessary here now.

Dimitri-Gnidash replied:
👍 The are both present.

tt-rkim commented on .github/workflows/docs-release.yaml:
```diff
@@ -0,0 +1,115 @@
+          echo "Copy tutorials and examples for TTNN from a different repo area."
+          cp -Lr ttnn/examples ttnn-docs/ttnn
+          echo "Make sure the links are updated in ttnn/usage.rst to the correct spot for the tutorials and examples."
+          sed -i "s/..\/..\/..\/..\/ttnn\/examples/examples/g" ttnn-docs/ttnn/usage.rst
```
This must have been horrible to figure out. Cheers to you!

tt-rkim commented on .github/workflows/docs-release.yaml:
```diff
@@ -0,0 +1,115 @@
+          echo "Register new documentation versions"
+          python update_tags.py tt-metalium $DOCS_VERSION
+          python update_tags.py ttnn $DOCS_VERSION
+          git checkout mains
```
`mains`?

Dimitri-Gnidash replied:
Thank you for catching this!

dimitri-tenstorrent committed:
#8534: Good eye Raymond! Fixinging the branch name.
